atomicfu/src/nativeMain/kotlin/kotlinx/atomicfu/Synchronized.kt

```diff
@@ -0,0 +1,204 @@
+                            mutexPool.release(mutex)
+                        }
+                    }
+                }
+                FAT -> {
+                    if (currentThreadId == state.ownerThreadId) {
+                        // nested lock
+                        val nestedFatLock = LockState(FAT, state.nestedLocks + 1, state.waiters, state.ownerThreadId, state.mutex)
+                        if (lock.compareAndSet(state, nestedFatLock)) return
+                    } else {
+                        val fatLock = LockState(FAT, state.nestedLocks, state.waiters + 1, state.ownerThreadId, state.mutex)
+                        if (lock.compareAndSet(state, fatLock)) {
+                            fatLock.mutex!!.lock()
+                            tryLockAfterResume(currentThreadId)
+                            return
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    public fun unlock() {
+        val currentThreadId = pthread_self()
+        while(true) {
+            val state = lock.value
+            require(currentThreadId == state.ownerThreadId) { "Thin lock may be only released by the owner thread" }
+            when(state.status) {
+                THIN -> {
+                    // nested unlock
+                    if (state.nestedLocks == 1) {
+                        val unlocked =  LockState(UNLOCKED, 0, 0)
+                        if (lock.compareAndSet(state, unlocked)) return
+                    } else {
+                        val releasedNestedLock = LockState(THIN, state.nestedLocks - 1, state.waiters, state.ownerThreadId)
+                        if (lock.compareAndSet(state, releasedNestedLock)) return
+                    }
+                }
+                FAT -> {
+                    if (state.nestedLocks == 1) {
+                        // last nested unlock -> release completely, resume some waiter
+                        val releasedLock = LockState(FAT, 0, state.waiters - 1, null, state.mutex)
+                        if (lock.compareAndSet(state, releasedLock)) {
+                            releasedLock.mutex!!.unlock()
+                            return
+                        }
+                    } else {
+                        // lock is still owned by the current thread
+                        val releasedLock = LockState(FAT, state.nestedLocks - 1, state.waiters, state.ownerThreadId, state.mutex)
+                        if (lock.compareAndSet(state, releasedLock)) return
+                    }
+                }
+                else -> error("It is not possible to unlock the mutex that is not obtained")
+            }
+        }
+    }
+
+    private fun tryLockAfterResume(threadId: pthread_t) {
+        while (true) {
+            val state = lock.value
+            if (state.ownerThreadId != null) continue
+            val mutex = state.mutex!!
+            val newState = if (state.waiters == 0) // deflate
```

Comment:
How can this `if (state.waiters == 0)` be ever true here? _This_ locker had incremented the waiters count so it looks like `assert(state.waiter >= 1)` can be safely added here.

Reply:
When some thread releases the lock, it sets `ownerThreadId` field to null, decrements the number of waiters and wakes up one of the waiting threads. So, here when the resumed thread tries to obtain the lock in `tryLockAfterResume` it is not accounted in the `waiters` and it deflates the lock if there are no more waiters.